client/lib/create-selector/index.js

```diff
@@ -31,38 +29,45 @@ const VALID_ARG_TYPES = [ 'number', 'boolean', 'string' ];
 const DEFAULT_GET_DEPENDANTS = state => state;
 
 /**
- * At runtime, assigns a function which returns a cache key for the memoized
- * selector function, given a state object and a variable set of arguments. In
- * development mode, this warns when the memoized selector is passed a complex
+ * In development mode, this warns when the memoized selector is passed a complex
  * object argument, as these cannot be depended upon as reliable cache keys.
+ * Note that the underlying function is created at runtime depending on the environment
  *
  * @type {Function} Function returning cache key for memoized selector
  */
-const DEFAULT_GET_CACHE_KEY = ( () => {
+const warnAgainstComplexArguments = ( () => {
 	let warn, includes;
 	if ( 'production' !== process.env.NODE_ENV ) {
 		// Webpack can optimize bundles if it can detect that a block will
 		// never be reached. Since `NODE_ENV` is defined using DefinePlugin,
 		// these debugging modules will be excluded from the production build.
 		warn = require( 'lib/warn' );
 		includes = require( 'lodash/includes' );
-	} else {
-		return ( state, ...args ) => args.join();
 	}
 
-	return ( state, ...args ) => {
+	return ( ...args ) => {
 		const hasInvalidArg = args.some( arg => {
 			return arg && ! includes( VALID_ARG_TYPES, typeof arg );
 		} );
 
 		if ( hasInvalidArg ) {
-			warn( 'Do not pass complex objects as arguments for a memoized selector' );
+			warn(
```

Comment:
I believe this would error in production? Since `warn` at this point would be `undefined` given the logic path above.

Also begs the question why we bother iterating over args if we don't anticipate doing anything in production. Seems the closure should just return `noop` in that environment.